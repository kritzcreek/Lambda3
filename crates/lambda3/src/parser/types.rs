use super::*;
use crate::lexer::SyntaxKind::*;

pub fn typ(p: &mut Parser) {
    let cp = p.checkpoint();
    if !atom(p) {
        p.report_error("expected type".to_string())
    }
    if p.at(ARROW) {
        p.finish_at(cp, TY_ARG);
        p.bump(ARROW);
        let cp2 = p.checkpoint();
        typ(p);
        p.finish_at(cp2, TY_RES);
        p.finish_at(cp, FUNC_TY)
    }
}

fn atom(p: &mut Parser) -> bool {
    match p.current() {
        INT => {
            p.start_node(INT_TY);
            p.bump(INT);
            p.finish_node();
        }
        BOOL => {
            p.start_node(BOOL_TY);
            p.bump(BOOL);
            p.finish_node()
        }
        L_PAREN => {
            p.start_node(PAREN_TY);
            p.bump(L_PAREN);
            typ(p);
            if !p.eat(R_PAREN) {
                p.report_error("missing closing paren".to_string())
            }
            p.finish_node()
        }
        _ => return false,
    }
    true
}
