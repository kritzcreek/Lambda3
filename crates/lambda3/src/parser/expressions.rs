use super::*;
use crate::lexer::SyntaxKind::*;

/// The outcome of parsing a single S-expression
pub enum ExprRes {
    /// An expression was successfully parsed
    Ok,
    /// An unexpected token was found
    Lul(String),
}

pub fn expr(p: &mut Parser) -> ExprRes {
    let mut is_application = false;
    let checkpoint = p.checkpoint();
    match atom(p) {
        None => return ExprRes::Lul("Expected expression".to_string()),
        Some(ExprRes::Ok) => (),
        Some(ExprRes::Lul(err)) => {
            p.errors.push(err.clone());
            return ExprRes::Lul(err);
        }
    }
    loop {
        let arg_checkpoint = p.checkpoint();
        match atom(p) {
            None => break,
            Some(ExprRes::Lul(s)) => {
                p.report_error(s);
            }
            Some(ExprRes::Ok) => {
                //TODO: something's broken
                p.finish_at(arg_checkpoint, EXPR_ARG);
                p.start_node_at(checkpoint, APPLICATION_E);
                is_application = true;
            }
        }
    }
    if is_application {
        p.finish_node();
    }
    ExprRes::Ok
}

// Option<ExprRes>
// Some(OK) => atom, continue
// Some(Err) => handle recovery
// None => no atom, no tokens

fn atom(p: &mut Parser) -> Option<ExprRes> {
    match p.current() {
        L_PAREN => parse_parenthesized(p),
        IDENT => parse_var(p),
        NUMBER_LIT | TRUE_KW | FALSE_KW => parse_literal(p),
        BACKSLASH => return parse_lambda(p),
        LET_KW => parse_let(p),
        LEXER_ERROR => p.bump_any(),
        _ => return None,
    }

    Some(ExprRes::Ok)
}

fn parse_let(p: &mut Parser) {
    p.start_node(LET_E);
    p.bump(LET_KW);
    patterns::parse_pattern(p);
    if !p.eat(EQUALS) {
        p.report_error(format!(
            "expected '=' in let expression but got {:?}",
            p.current()
        ))
    }
    if let ExprRes::Lul(err) = expr(p) {
        p.start_node(ERROR);
        p.errors.push(err);
        while !p.eat(IN_KW) {}
        p.finish_node();
    }

    if !p.eat(IN_KW) {
        p.report_error(format!(
            "expected 'in' in let expression but got {:?}",
            p.current()
        ))
    }

    p.start_node(EXPR_LET_BODY);
    if let ExprRes::Lul(err) = expr(p) {
        p.report_error(err)
    };
    p.finish_node();

    p.finish_node();
}

fn parse_parenthesized(p: &mut Parser) {
    p.start_node(PAREN_E);
    p.bump(L_PAREN);
    expr(p);

    if !p.eat(R_PAREN) {
        p.start_node(ERROR);
        p.errors
            .push(format!("unexpected token {:?}, expected ')'", p.current()));

        while !(p.at(R_PAREN) || p.at(EOF)) {
            p.bump_any()
        }

        p.finish_node();
        p.eat(R_PAREN);
    }

    p.finish_node();
}

fn parse_var(p: &mut Parser) {
    p.start_node(VAR_E);
    p.bump(IDENT);
    p.finish_node();
}

fn parse_literal(p: &mut Parser) {
    p.start_node(LITERAL_E);
    p.bump_any();
    p.finish_node();
}

fn parse_lambda(p: &mut Parser) -> Option<ExprRes> {
    p.start_node(LAMBDA_E);
    p.bump(BACKSLASH);

    patterns::parse_pattern(p);

    if !p.eat(DOT) {
        p.finish_node();
        return Some(ExprRes::Lul(format!("expected '.', got {:?}", p.current())));
    }

    match expr(p) {
        ExprRes::Ok => p.finish_node(),
        ExprRes::Lul(err) => {
            p.report_error(err);
            p.finish_node();
            return None;
        }
    }

    Some(ExprRes::Ok)
}
