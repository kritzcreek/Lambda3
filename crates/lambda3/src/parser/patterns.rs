use super::*;
use crate::lexer::SyntaxKind::*;

pub fn parse_pattern(p: &mut Parser) {
    let checkpoint = p.checkpoint();

    let current = p.current();
    match current {
        IDENT => {
            p.bump(IDENT);
            p.finish_at(checkpoint, VAR_P);
        }
        UNDERSCORE => {
            p.bump(UNDERSCORE);
            p.finish_at(checkpoint, WILDCARD_P);
        }
        L_PAREN => {
            p.bump(L_PAREN);
            parse_pattern(p);
            if !p.eat(R_PAREN) {
                p.report_error(format!("missing closing paren, got {:?}", p.current()));
            }
            p.finish_at(checkpoint, PAREN_P);
        }
        _ => {
            p.report_error(format!("expected pattern, got {:?}", current));
            return;
        }
    }

    opt_annotation(p, checkpoint);
}

fn opt_annotation(p: &mut Parser, checkpoint: Checkpoint) -> bool {
    if p.eat(COLON) {
        types::typ(p);
        p.finish_at(checkpoint, ANNOTATION_P);
        return true;
    }

    false
}
