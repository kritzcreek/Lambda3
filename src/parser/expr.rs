use super::*;
use crate::lexer::SyntaxKind::*;

/// The outcome of parsing a single S-expression
pub enum ExprRes {
    /// An expression was successfully parsed
    Ok,
    /// An unexpected token was found
    Lul(String),
}

pub fn expr(p: &mut Parser) -> ExprRes {
    let mut is_application = false;
    let checkpoint = p.checkpoint();
    match atom(p) {
        None => return ExprRes::Lul("Expected expression".to_string()),
        Some(ExprRes::Ok) => (),
        Some(ExprRes::Lul(err)) => {
            p.errors.push(err.clone());
            return ExprRes::Lul(err);
        }
    }
    loop {
        match atom(p) {
            None => break,
            Some(ExprRes::Lul(s)) => {
                p.report_error(s);
            }
            Some(ExprRes::Ok) => {
                p.start_node_at(checkpoint, ApplicationE);
                is_application = true;
            }
        }
    }
    if is_application {
        p.finish_node();
    }
    ExprRes::Ok
}

// Option<ExprRes>
// Some(OK) => atom, continue
// Some(Err) => handle recovery
// None => no atom, no tokens

fn atom(p: &mut Parser) -> Option<ExprRes> {
    match p.current() {
        LParen => parse_parenthesized(p),
        Ident => parse_var(p),
        NumberLit | TrueKw | FalseKw => parse_literal(p),
        Backslash => return parse_lambda(p),
        LexerError => p.bump_any(),
        _ => return None,
    }

    Some(ExprRes::Ok)
}

fn parse_parenthesized(p: &mut Parser) {
    p.start_node(ParenE);
    p.bump(LParen);
    expr(p);

    if !p.eat(RParen) {
        p.start_node(Error);
        p.errors.push(format!(
            "unexpected token {:?}, expected ')'",
            p.current()
        ));

        while p.current() != RParen && p.current() != Eof {
            p.bump_any()
        }

        p.finish_node();
        p.eat(RParen);
    }

    p.finish_node();
}

fn parse_var(p: &mut Parser) {
    p.start_node(VarE);
    p.bump(Ident);
    p.finish_node();
}

fn parse_literal(p: &mut Parser) {
    p.start_node(LiteralE);
    p.bump_any();
    p.finish_node();
}

fn parse_lambda(p: &mut Parser) -> Option<ExprRes> {
    p.start_node(LambdaE);
    p.bump(Backslash);

    let checkpoint = p.checkpoint();
    if p.eat(Ident) {
        p.start_node_at(checkpoint, VarP);
        p.finish_node();
    } else {
        p.report_error(format!("expected binder, got {:?}", p.current()))
    }

    if p.eat(Colon) {
        types::typ(p)
    }

    if !p.eat(Dot) {
        p.finish_node();
        return Some(ExprRes::Lul(format!(
            "expected '.', got {:?}",
            p.current()
        )));
    }

    match expr(p) {
        ExprRes::Ok => p.finish_node(),
        ExprRes::Lul(err) => {
            p.report_error(err);
            p.finish_node();
            return None;
        }
    }

    Some(ExprRes::Ok)
}

