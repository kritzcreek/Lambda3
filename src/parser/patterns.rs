use crate::lexer::SyntaxKind::*;
use super::*;

pub fn parse_pattern(p: &mut Parser) {
    let checkpoint = p.checkpoint();

    let current = p.current();
    match current {
        Ident => {
            p.bump(Ident);
            p.finish_at(checkpoint, VarP);
        }
        Underscore => {
            p.bump(Underscore);
            p.finish_at(checkpoint, WildcardP);
        }
        LParen => {
            p.bump(LParen);
            parse_pattern(p);
            if !p.eat(RParen) {
                p.report_error(format!("missing closing paren, got {:?}", p.current()));
            }
            p.finish_at(checkpoint, ParenP);
        }
        _ => {
            p.report_error(format!("expected pattern, got {:?}", current));
            return;
        }
    }

    opt_annotation(p, checkpoint);
}

fn opt_annotation(p: &mut Parser, checkpoint: Checkpoint) -> bool {
    if p.eat(Colon) {
        types::typ(p);
        p.finish_at(checkpoint, AnnotationP);
        return true;
    }

    return false;
}