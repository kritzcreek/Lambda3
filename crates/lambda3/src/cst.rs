use crate::parser::{self, Parse};

pub fn parse(text: &str) -> Parse {
    parser::parse(text)
}

pub fn parse_type(text: &str) -> Parse {
    parser::parse_type(text)
}

pub fn parse_pattern(text: &str) -> Parse {
    parser::parse_pattern(text)
}

#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, glob};

    use super::*;

    #[test]
    fn parse_expr_test() {
        use std::fs;

        glob!("examples/expr/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }

    #[test]
    fn parse_type_test() {
        use std::fs;

        glob!("examples/types/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse_type(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }

    #[test]
    fn parse_patterns_test() {
        use std::fs;

        glob!("examples/patterns/*.l3", |path| {
            let input = fs::read_to_string(path).unwrap();
            let parse = parse_pattern(&input);
            let node = parse.syntax();

            assert_debug_snapshot!((input, parse.errors, node));
        });
    }
}
