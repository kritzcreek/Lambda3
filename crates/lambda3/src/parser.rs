use rowan::{Checkpoint, GreenNode, GreenNodeBuilder, Language};

use crate::cst;
use crate::lexer::SyntaxKind::{EOF, ERROR, ROOT};
use crate::lexer::{lex_str, SyntaxKind, Token};
use crate::syntax::ast::AstNode;
use crate::syntax::nodes;
use crate::syntax::{Lang, SyntaxNode};

mod expressions;
mod patterns;
mod types;

/// The parse results are stored as a "green tree".
/// We'll discuss working with the results later
#[derive(Debug)]
pub struct Parse {
    pub green_node: GreenNode,
    #[allow(unused)]
    pub errors: Vec<String>,
}

pub struct Parser {
    /// input tokens, including whitespace,
    /// in *reverse* order.
    tokens: Vec<(Vec<Token>, Token)>,
    /// the in-progress tree.
    builder: GreenNodeBuilder<'static>,
    /// the list of syntax errors we've accumulated
    /// so far.
    errors: Vec<String>,
}

impl Parse {
    pub fn syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green_node.clone())
    }

    pub fn expr(&self) -> nodes::Expr {
        self.syntax()
            .children()
            .find_map(|n| nodes::Expr::cast(n))
            .unwrap()
    }

    pub fn debug_tree(&self) -> String {
        let syntax_node = SyntaxNode::new_root(self.green_node.clone());
        let formatted = format!("{:#?}", syntax_node);

        // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
        formatted[0..formatted.len() - 1].to_string()
    }
}

impl Parser {
    fn new(tokens: Vec<(Vec<Token>, Token)>) -> Self {
        Parser {
            tokens,
            builder: GreenNodeBuilder::new(),
            errors: Vec::new(),
        }
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        self.builder.start_node(Lang::kind_to_raw(kind));
    }

    fn start_node_at(&mut self, checkpoint: Checkpoint, kind: SyntaxKind) {
        self.builder
            .start_node_at(checkpoint, Lang::kind_to_raw(kind));
    }

    fn finish_node(&mut self) {
        self.builder.finish_node();
    }

    fn finish_at(&mut self, c: Checkpoint, kind: SyntaxKind) {
        self.start_node_at(c, kind);
        self.finish_node();
    }

    fn checkpoint(&self) -> Checkpoint {
        self.builder.checkpoint()
    }

    fn report_error(&mut self, msg: String) {
        self.start_node(ERROR);
        self.errors.push(msg);
        self.finish_node();
    }

    fn nth(&self, n: usize) -> SyntaxKind {
        let len = self.tokens.len();
        if n >= len {
            SyntaxKind::EOF
        } else {
            self.tokens[len - n - 1].1 .0
        }
    }

    /// Peek at the first unprocessed token
    fn current(&self) -> SyntaxKind {
        self.nth(0)
    }

    fn at(&self, kind: SyntaxKind) -> bool {
        self.current() == kind
    }

    fn eat(&mut self, kind: SyntaxKind) -> bool {
        if !self.at(kind) {
            return false;
        }
        self.bump_any();
        true
    }

    /// Advance one token, adding it to the current branch of the tree builder.
    fn bump_any(&mut self) {
        if self.at(EOF) {
            return;
        }
        let (leading, (kind, text)) = self.tokens.pop().unwrap();

        for (kind, text) in leading {
            self.builder.token(Lang::kind_to_raw(kind), text.as_str());
        }

        self.builder.token(Lang::kind_to_raw(kind), text.as_str());
    }

    fn bump(&mut self, kind: SyntaxKind) {
        assert!(self.eat(kind))
    }
}

pub fn parse(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(ROOT);
    let _ = expressions::expr(&mut p);
    if !p.at(SyntaxKind::EOF) {
        p.start_node(ERROR);
        while !p.at(EOF) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}

pub fn parse_type(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(ROOT);
    let _ = types::typ(&mut p);
    if !p.at(SyntaxKind::EOF) {
        p.start_node(ERROR);
        while !p.at(EOF) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}

pub fn parse_pattern(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(ROOT);
    let _ = patterns::parse_pattern(&mut p);
    if !p.at(SyntaxKind::EOF) {
        p.start_node(ERROR);
        while !p.at(EOF) {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors,
    }
}
