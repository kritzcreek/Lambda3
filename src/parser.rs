use rowan::{Checkpoint, GreenNode, GreenNodeBuilder, Language, SmolStr};

use crate::cst::Root;
use crate::lexer::SyntaxKind::{Eof, Error, RootE, Whitespace};
use crate::lexer::{lex_str, SyntaxKind};
use crate::syntax::{Lang, SyntaxNode};

mod expr;

/// The parse results are stored as a "green tree".
/// We'll discuss working with the results later
#[derive(Debug)]
pub struct Parse {
    pub green_node: GreenNode,
    #[allow(unused)]
    pub errors: Vec<String>,
}

pub struct Parser {
    /// input tokens, including whitespace,
    /// in *reverse* order.
    tokens: Vec<(Vec<(SyntaxKind, SmolStr)>, (SyntaxKind, SmolStr))>,
    /// the in-progress tree.
    builder: GreenNodeBuilder<'static>,
    /// the list of syntax errors we've accumulated
    /// so far.
    errors: Vec<String>,
}

impl Parse {
    pub fn syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green_node.clone())
    }

    pub fn root(&self) -> Root {
        Root::cast(self.syntax()).unwrap()
    }

    pub fn debug_tree(&self) -> String {
        let syntax_node = SyntaxNode::new_root(self.green_node.clone());
        let formatted = format!("{:#?}", syntax_node);

        // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
        formatted[0..formatted.len() - 1].to_string()
    }
}

impl Parser {
    fn new(tokens: Vec<(Vec<(SyntaxKind, SmolStr)>, (SyntaxKind, SmolStr))>) -> Self {
        Parser {
            tokens,
            builder: GreenNodeBuilder::new(),
            errors: Vec::new(),
        }
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        self.builder.start_node(Lang::kind_to_raw(kind));
    }

    fn start_node_at(&mut self, checkpoint: Checkpoint, kind: SyntaxKind) {
        self.builder
            .start_node_at(checkpoint, Lang::kind_to_raw(kind));
    }

    fn finish_node(&mut self) {
        self.builder.finish_node();
    }

    fn finish_at(&mut self, c: Checkpoint, kind: SyntaxKind) {
        self.start_node_at(c, kind);
        self.finish_node();
    }

    fn checkpoint(&self) -> Checkpoint {
        self.builder.checkpoint()
    }

    fn report_error(&mut self, msg: String) {
        self.start_node(Error);
        self.errors.push(msg);
        self.finish_node();
    }

    fn nth(&self, n: usize) -> SyntaxKind {
        let len = self.tokens.len();
        if n >= len {
            SyntaxKind::Eof
        } else {
            self.tokens[len - n - 1].1 .0
        }
    }

    /// Peek at the first unprocessed token
    fn current(&self) -> SyntaxKind {
        self.nth(0)
    }

    fn nth_at(&self, n: usize, kind: SyntaxKind) -> bool {
        self.nth(n) == kind
    }

    fn at(&self, kind: SyntaxKind) -> bool {
        self.current() == kind
    }

    fn eat(&mut self, kind: SyntaxKind) -> bool {
        if self.current() != kind {
            return false;
        }
        self.bump_any();
        true
    }

    /// Advance one token, adding it to the current branch of the tree builder.
    fn bump_any(&mut self) {
        if self.current() == Eof {
            return;
        }
        let (leading, (kind, text)) = self.tokens.pop().unwrap();

        for (kind, text) in leading {
            self.builder.token(Lang::kind_to_raw(kind), text);
        }

        self.builder.token(Lang::kind_to_raw(kind), text);
    }

    fn bump(&mut self, kind: SyntaxKind) {
        assert!(self.eat(kind))
    }
}

pub fn parse(text: &str) -> Parse {
    let mut tokens = lex_str(text);
    tokens.reverse();
    let mut p = Parser::new(tokens);

    p.start_node(RootE);
    let _ = expr::expr(&mut p);
    if p.current() != SyntaxKind::Eof {
        p.start_node(Error);
        while p.current() != Eof {
            p.bump_any()
        }
        p.finish_node();
        p.errors.push(format!("Unexpected token {:?}", p.current()))
    }
    p.finish_node();

    Parse {
        green_node: p.builder.finish(),
        errors: p.errors.clone(),
    }
}
