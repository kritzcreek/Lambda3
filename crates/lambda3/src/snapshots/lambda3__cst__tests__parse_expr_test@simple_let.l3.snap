---
source: crates/lambda3/src/cst.rs
expression: "(input, parse.errors, node)"
input_file: crates/lambda3/src/examples/expr/simple_let.l3

---
(
    "let _ = 1 in 2",
    [],
    ROOT@0..14
      LET_E@0..14
        LET_KW@0..3 "let"
        WILDCARD_P@3..5
          WHITESPACE@3..4 " "
          UNDERSCORE@4..5 "_"
        WHITESPACE@5..6 " "
        EQUALS@6..7 "="
        LITERAL_E@7..9
          WHITESPACE@7..8 " "
          NUMBER_LIT@8..9 "1"
        WHITESPACE@9..10 " "
        IN_KW@10..12 "in"
        EXPR_LET_BODY@12..14
          LITERAL_E@12..14
            WHITESPACE@12..13 " "
            NUMBER_LIT@13..14 "2"
    ,
)
